File: algorithms/cpp/maximumNumberofNon-OverlappingSubstrings/maximumNumberofNon-OverlappingSubstrings.cpp
// Source: https://leetcode.com/problems/maximum-number-of-non-overlapping-substrings/
// Author: Miao Zhang
// Date:   2021-05-13

class Solution {
public:
    vector<string> maxNumOfSubstrings(string s) {
        int n = s.length();
        vector<int> l(26, INT_MAX);
        vector<int> r(26, INT_MIN);
        for (int i = 0; i < n; i++) {
            l[s[i] - 'a'] = min(l[s[i] - 'a'], i);
            r[s[i] - 'a'] = max(r[s[i] - 'a'], i);
        }
        auto extend = [&] (int i) -> int {
            int p = r[s[i] - 'a'];
            for (int j = i; j <= p; j++) {
                if (l[s[j] - 'a'] < i)
                    return -1;
                p = max(p, r[s[j] - 'a']);
            }
            return p;
        };
        vector<string> res;
        int last = -1;
        for (int i = 0; i < n; i++) {
            if (i != l[s[i] - 'a']) continue;
            int p = extend(i);
            if (p == -1) continue;
            if (i > last) res.push_back("");
            res.back() = s.substr(i, p - i + 1);
            last = p;
        }
        return res;
    }
};

File: algorithms/python/maximumNumberofNon-OverlappingSubstrings/maximumNumberofNon-OverlappingSubstrings.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/maximum-number-of-non-overlapping-substrings/
# Author: Miao Zhang
# Date:   2021-05-13

class Solution:
    def maxNumOfSubstrings(self, s: str) -> List[str]:
        n = len(s)
        l = [float('inf')] * 26
        r = [float('-inf')] * 26
        for i in range(n):
            l[ord(s[i]) - ord('a')] = min(l[ord(s[i]) - ord('a')], i)
            r[ord(s[i]) - ord('a')] = max(r[ord(s[i]) - ord('a')], i)
        
        def extend(i):
            p = r[ord(s[i]) - ord('a')]
            j = i
            while j <= p:
                if l[ord(s[j]) - ord('a')] < i:
                    return -1
                p = max(p, r[ord(s[j]) - ord('a')])
                j += 1
            return p
        
        res = []
        last = -1
        for i in range(n):
            if i != l[ord(s[i]) - ord('a')]:
                continue
            p = extend(i)
            if p == -1: continue
            if i > last:
                res.append('')
            res[-1] = s[i: p + 1]
            last = p
        return res
